text/000-rustdoc-bundle-local-resources.md

```diff
@@ -0,0 +1,118 @@
+
+```rust
+/// Using a local image
+#[doc=concat!("![with absolute path](", env!("OUT_DIR"), "/local/image.png)")]
+///
+/// Using a local image ![with relative path](../local/image.png)
+```
+
+Since the local images are all put in the same folder, if the same is imported from different crates, the content won't be duplicated since they have the same name and the same hash.
+
+If the path isn't referring to a file, a warning will be emitted and rustdoc will left the path unchanged in the generated documentation.
+
+For published crates, `docs.rs` builds the contents of the `.crate` package in a sandbox with no internet access. Make sure any resources your docs need are [included](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) in the package.
+
+The local resources files are not affected by the `--resource-suffix`.
+
+The impact on `docs.rs` would also be very minimal as the size of a published crate resources is limited to a few megabytes. The only thing needed would be to handle the new `doc.files` folder.
+
+There are two options that will be impacted by this RFC: the favicon and logo that you can respectively set through:
+
+```rust
+#![doc(html_favicon_url = "some_path", html_logo_url = "some_other_path")]
+```
+
+They will follow the same rule as for other images: if this is a local path, the local file will be copied and the paths to it will be rewritten.
+
+To support `#[doc(inline)]` for foreign items using local resources, it will rely on the `-Zrustdoc-map` option.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new rustdoc pass will be added which would go through all documentation to gather local resources into a map.
+
+Then in HTML documentation generation, the local resources pathes will be replaced by their equivalent linking to the output directory instead.
+
+The local resources files will be renamed as follows: `{original filename}-{hash}{extension}`. The `{hash}` information will be computed from the local resource file content.
+
+You can look at what the implementation could look like in [#107640](https://github.com/rust-lang/rust/pull/107640).
+
+When an image is included in an item that gets inlined across a crate, rustdoc will treat it like a cross-crate intra-doc link, using `--extern-html-root-url` so that `docs.rs` can hotlink the image from the crate that holds a copy of the image. This has a few upsides and downsides compared to the approach where the image itself is copied into the crate with the inlined docs.
+
+ * reduces the number of duplicated images that `docs.rs` has to store
+ * doesn't require the source code for the source crate when inlining
+ * only requires storing the hash of the file in the `.rmeta`, not the whole image
+ * requires rustc to look at the doc comments and hash the image(s)
```

Comment:
Perhaps; or perhaps the lint should be something like "the total of all your non-code resources is >10% of the crate size." It's still not really satisfying though, since it doesn't fully solve the problem. If you want to include lots of resources, or large resources, the only solution is going to be to refer to them using absolute URLs.

Out of curiosity I spot-checked a couple of crates, [regex](https://crates.io/crates/regex) and [rayon](https://crates.io/crates/rayon). They're 248 kB and 169 kB respectively, so 10% would be 24.8 kB and 16.9 kB.

Also it's worth mentioning one counterpoint to my concern: source files that are bundled into .crate files are not minified and comments aren't removed. That's important specifically so docs can be generated from them (including the source file view within rustdoc). So in some sense published crates already do contain bytes that are only useful for documentation.

Reply:
> So in some sense published crates already do contain bytes that are only useful for documentation.

Examples and tests are also arguably superfluous files — they benefit [crater](https://github.com/rust-lang/crater), and those who read the source code of the downloaded crate, but not the typical downloader.

Reply:
Random musings on cargo's behavior

To protect against zip bombs, cargo had a 512 MB limit when extracting crates and people ran into this (rust-lang/cargo#11151) and we updated it to take compression size into account (rust-lang/cargo#11337) so we are less likely to hit the 512 MB limit unintentionally.

We also now report crate size on package/publish (rust-lang/cargo#11270).  One future possibility mentioned in that PR is to warn when binary files are included (rust-lang/cargo#9058).  In general, cargo has been hesitant about adding warnings that because we haven't had a way for people to disable them but with `[lints]` (rust-lang/cargo#12115), we have the possibility to add `[lints.cargo]` support.

Reply:
We should be cognizant of users.  By including files that aren't as necessary
- Slower download of `.crate` file
- Extra disk space taken up (currently will be twice, once for the `.crate` and once for the extracted directory)
- Extra time to extract the `.crate`

Some things that would help (and are [currently being discussed](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/last-use.20tracking))
- Garbage collection of `.crate` files and extracted `.crate` files
- Building directly from `.crate` files (this is a big more out there)

Dependent CI jobs will never use this content.

I wonder how many users are like me that almost exclusively use docs.rs rather than `cargo doc` and any missing images wouldn't be a big deal.  Or in other words, we should consider if the benefit to the community outweighs the cost to the community.

Reply:
Just answering:

> I wonder how many users are like me that almost exclusively use docs.rs rather than `cargo doc` and any missing images wouldn't be a big deal. Or in other words, we should consider if the benefit to the community outweighs the cost to the community.

It'd be needed for docs.rs too unfortunately.

Reply:
> It'd be needed for docs.rs too unfortunately.

How so?  I'm not seeing anything listed in the Motivation that benefits us docs.rs-exclusive users (speaking as someone who also has crates that include images).